File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/gemstone/StreamConnection.kt
package com.gemwallet.android.data.repositories.gemstone

import com.gemwallet.android.data.repositories.stream.WebSocketConnectable
import uniffi.gemstone.GemStreamConnection

class GemstoneStreamConnection(
    private val connection: WebSocketConnectable,
) : GemStreamConnection {
    override suspend fun isConnected(): Boolean = connection.isConnected

    override suspend fun send(message: String) {
        check(connection.send(message)) { "Stream connection is closed" }
    }
}

File: ios/Packages/FeatureServices/StreamService/StreamConnection.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import protocol Gemstone.GemStreamConnection
import WebSocketClient

public final class GemstoneStreamConnection: GemStreamConnection, Sendable {
    private let webSocket: any WebSocketConnectable

    public init(webSocket: any WebSocketConnectable) {
        self.webSocket = webSocket
    }

    public func isConnected() async -> Bool {
        await webSocket.state == .connected
    }

    public func send(message: String) async throws {
        try await webSocket.send(message)
    }
}
